### ISSUE TITLE
uni import: tail calls must be recognized and handled, but where?

### ISSUE THREAD

**Issue by matsc-at-sics-se** [matsc-at-sics-se/unison#10]
[spio_mutex_pthread.zip](https://github.com/matsc-at-sics-se/unison/files/1780640/spio_mutex_pthread.zip)
_Reproducer:_

```
$ uni import -f spio_mutex_unlock_pthread spio_mutex_pthread.mir -o spio_mutex_unlock_pthread.uni --target=X86 --maxblocksize=50 --goal=speed --copyremat -d
```

_(gdb) disass spio_mutex_unlock_pthread_
```
Dump of assembler code for function spio_mutex_unlock_pthread:
   0x00007ffff7b66730 <+0>:	sub    $0x8,%rsp
   0x00007ffff7b66734 <+4>:	callq  0x7ffff7aac2a0 <pthread_mutex_unlock@plt>
   0x00007ffff7b66739 <+9>:	test   %eax,%eax
   0x00007ffff7b6673b <+11>:	jne    0x7ffff7b66748 <spio_mutex_unlock_pthread+24>
   0x00007ffff7b6673d <+13>:	xor    %eax,%eax
   0x00007ffff7b6673f <+15>:	add    $0x8,%rsp
   0x00007ffff7b66743 <+19>:	retq   
   0x00007ffff7b66744 <+20>:	nopl   0x0(%rax)
   0x00007ffff7b66748 <+24>:	mov    $0xffffd8df,%esi
   0x00007ffff7b6674d <+29>:	mov    %eax,%edi
   0x00007ffff7b6674f <+31>:	add    $0x8,%rsp
   0x00007ffff7b66753 <+35>:	jmpq   0x7ffff7b65ec0 <spio_map_unix_error>
```

_*.mir, IR part_
```
  define i32 @spio_mutex_unlock_pthread(%struct.spio_t_mutex_* %mutex) #0 {
    %1 = bitcast %struct.spio_t_mutex_* %mutex to %union.pthread_mutex_t*
    %2 = tail call i32 @pthread_mutex_unlock(%union.pthread_mutex_t* %1) #4
    %3 = icmp eq i32 %2, 0
    br i1 %3, label %6, label %4
  
  ; <label>:4                                       ; preds = %0
    %5 = tail call i32 @spio_map_unix_error(i32 %2, i32 -10017) #4
    ret i32 %5
  
  ; <label>:6                                       ; preds = %0
    ret i32 0
  }
```

_*.mir, MIR part_
```
body:             |
  bb.0 (%ir-block.0, freq 21):
    successors: %bb.2(37), %bb.1(62)
    liveins: %rdi
  
    %1 = COPY %rdi
    ADJCALLSTACKDOWN64 0, 0, implicit-def dead %rsp, implicit-def dead %eflags, implicit %rsp
    %rdi = COPY %1
    CALL64pcrel32 @pthread_mutex_unlock, csr_64, implicit %rsp, implicit %rdi, implicit-def %rsp, implicit-def %eax
    ADJCALLSTACKUP64 0, 0, implicit-def dead %rsp, implicit-def dead %eflags, implicit %rsp
    %2 = COPY %eax
    TEST32rr %2, %2, implicit-def %eflags
    JE_1 %bb.2, implicit %eflags
    JMP_1 %bb.1
  
  bb.1 (%ir-block.4, freq 13):
    liveouts: %rsp, %edi, %esi
  
    %0 = COPY %2
    %3 = MOV32ri -10017
    %edi = COPY %0
    %esi = COPY %3
    TCRETURNdi64 @spio_map_unix_error, 0, csr_64, implicit %rsp, implicit %edi, implicit %esi
  
  bb.2 (%ir-block.6, freq 8):
    liveouts:
  
    %4 = MOV32r0 implicit-def dead %eflags
    %eax = COPY %4
    RETQ %eax
```

_*.uni, output from dropDebugLocations, normalizePhis, liftBlockFreqs, liftCustomProperties, liftBranchPredictions_
```
body:             |
  bb.0 (freq 21):
    successors: %bb.2(37), %bb.1(62)
  
    ENTRY %rdi
    COPY %1, %rdi
    ADJCALLSTACKDOWN64 0, 0, %rsp, %eflags, %rsp
    COPY %rdi, %1
    CALL64pcrel32 @pthread_mutex_unlock, csr_64, %rsp, %rdi, %rsp, %eax
    ADJCALLSTACKUP64 0, 0, %rsp, %eflags, %rsp
    COPY %2, %eax
    TEST32rr %2, %2, %eflags
    JE_1 %bb.2, %eflags
    JMP_1 %bb.1
  
  bb.1 (freq 13):
  
    COPY %0, %2
    MOV32ri %3, -10017
    COPY %edi, %0
    COPY %esi, %3
    TCRETURNdi64 @spio_map_unix_error, 0, csr_64, %rsp, %edi, %esi
    RETURN %rsp, %edi, %esi <<< BOGUS OPERATION AND OPERANDS
  
  bb.2 (freq 8):
  
    MOV32r0 %4, %eflags
    COPY %eax, %4
    RETQ %eax
    RETURN
```

_*.uni, output from liftJumpTables, liftMemoryPartitions, simplifyFallthroughs, renameMachineBlocks, splitTerminators, renameMachineBlocks, dropUnsupportedPseudos, liftMachineUndefs, extractSubRegs, lowerInsertSubRegs, lowerSubRegVirtuals, runPreProcess_
```
body:             |
  bb.0 (freq 21):
  
    ENTRY %rdi
    COPY %1, %rdi
    ADJCALLSTACKDOWN64 0, 0, %rsp, %eflags, %rsp
    COPY %rdi, %1
    CALL64pcrel32 @pthread_mutex_unlock, csr_64, %rsp, %rdi, %rsp, %eax
    ADJCALLSTACKUP64 0, 0, %rsp, %eflags, %rsp
    COPY %2, %eax
    TEST32rr %2, %2, %eflags
    JE_1 %bb.2, %eflags
  
  bb.1 (freq 13):
  
    COPY %0, %2
    MOV32ri %3, -10017
    COPY %edi, %0
    COPY %esi, %3
    TCRETURNdi64 @spio_map_unix_error, 0, csr_64, %rsp, %edi, %esi
    RETURN %rsp, %edi, %esi <<< BOGUS OPERATION AND OPERANDS
  
  bb.2 (freq 8):
  
    MOV32r0 %4, %eflags
    COPY %eax, %4
    RETQ %eax
    RETURN
```

_*.uni, output from liftGoal_
```
b0 (freq: 21):
    o0: [rdi] <- (entry) []
    o1: [t1] <- (copy) [rdi]
    o2: [] <- ADJCALLSTACKDOWN64 [0,0]
    o3: [rdi] <- (copy) [t1]
    o4: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [t2] <- (copy) [eax]
    o7: [] <- TEST32rr [t2,t2] <<< DEF OPERAND MISSING, CONGRUENT WITH t2
    o8: [] <- JE_1 [b2] (taken: false)
b1 (freq: 13):
    o9: [t0] <- (copy) [t2]
    o10: [t3] <- MOV32ri [-10017]
    o11: [edi] <- (copy) [t0]
    o12: [esi] <- (copy) [t3]
    o13: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o14: [] <- (return) [rsp,edi,esi] <<< BOGUS OPERATION AND OPERANDS
b2 (freq: 8):
    o15: [t4] <- MOV32r0 []
    o16: [eax] <- (copy) [t4]
    o17: [] <- RETQ [eax]
    o18: [] <- (return) []
```

_*.uni, output from addDelimiters, postponeBranches, removeUnreachableBlocks, renameBlocks, correctDoubleBranches, adjustPhiLabels, simplifyCombinations, removeUselessVirtuals, relocateDefines, runTargetTransforms_
```
b0 (entry, freq: 21):
    o0: [rdi] <- (in) []
    o1: [t1] <- (copy) [rdi]
    o2: [] <- ADJCALLSTACKDOWN64 [0,0]
    o3: [rdi] <- (copy) [t1]
    o4: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [t2] <- (copy) [eax]
    o7: [] <- TEST32rr [t2,t2] <<< DEF OPERAND MISSING, CONGRUENT WITH t2
    o8: [] <- JE_1 [b2] (taken: false)
    o19: [] <- (out) []
b1 (exit, return, freq: 13):
    o23: [] <- (in) []
    o9: [t0] <- (copy) [t2]
    o10: [t3] <- MOV32ri [-10017]
    o11: [edi] <- (copy) [t0]
    o12: [esi] <- (copy) [t3]
    o13: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o20: [] <- (out) [rsp,edi,esi] <<< BOGUS OPERANDS
b2 (exit, return, freq: 8):
    o24: [] <- (in) []
    o15: [t4] <- MOV32r0 []
    o16: [eax] <- (copy) [t4]
    o17: [] <- RETQ [eax]
    o21: [] <- (out) []
```

_*.uni, output from runTargetTransforms, extractCallRegs_
```
b0 (entry, freq: 21):
    o0: [rdi] <- (in) []
    o1: [t1] <- (copy) [rdi]
    o2: [] <- ADJCALLSTACKDOWN64 [0,0]
    o3: [rdi] <- (copy) [t1]
    o4: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [t2] <- (copy) [eax]
    o7: [] <- TEST32rr [t2,t2] <<< DEF OPERAND MISSING, CONGRUENT WITH t2
    o8: [] <- JE_1 [b2] (taken: false)
    o19: [] <- (out) []
b1 (exit, return, freq: 13):
    o23: [] <- (in) []
    o9: [t0] <- (copy) [t2]
    o10: [t3] <- MOV32ri [-10017]
    o11: [edi] <- (copy) [t0]
    o12: [esi] <- (copy) [t3]
    o13: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o20: [] <- (out) [rsp,edi,esi] <<< BOGUS OPERANDS
b2 (exit, return, freq: 8):
    o24: [] <- (in) []
    o15: [t4] <- MOV32r0 []
    o16: [eax] <- (copy) [t4]
    o17: [] <- RETQ [42]
    o21: [] <- (out) [t4:eax]
```

_*.uni, output from extractCallRegs_
```
b0 (entry, freq: 21):
    o0: [rdi] <- (in) []
    o1: [t1] <- (copy) [rdi]
    o2: [] <- ADJCALLSTACKDOWN64 [0,0]
    o3: [t5] <- (copy) [t1]
    o4: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o25: [t6:eax] <- (fun) [t5:rdi] (call: o4)
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [t2] <- (copy) [t6]
    o7: [] <- TEST32rr [t2,t2] <<< DEF OPERAND MISSING, CONGRUENT WITH t2
    o8: [] <- JE_1 [b2] (taken: false)
    o19: [] <- (out) []
b1 (exit, return, freq: 13):
    o23: [] <- (in) []
    o9: [t0] <- (copy) [t2]
    o10: [t3] <- MOV32ri [-10017]
    o11: [t7] <- (copy) [t0]
    o12: [t8] <- (copy) [t3]
    o13: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o26: [] <- (fun) [t7:edi,t8:esi] (call: o13)
    o20: [] <- (out) [rsp,edi,esi] <<< BOGUS OPERANDS
b2 (exit, return, freq: 8):
    o24: [] <- (in) []
    o15: [t4] <- MOV32r0 []
    o17: [] <- RETQ [42]
    o21: [] <- (out) [t4:eax]
```

_*.uni, output from liftRegs, runTargetTransforms_
```
b0 (entry, freq: 21):
    o0: [t9:rdi] <- (in) []
    o1: [t1] <- (copy) [t9]
    o2: [] <- ADJCALLSTACKDOWN64 [0,0]
    o3: [t5] <- (copy) [t1]
    o4: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o25: [t6:eax] <- (fun) [t5:rdi] (call: o4)
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [t2] <- (copy) [t6]
    o7: [] <- TEST32rr [t2,t2] <<< DEF OPERAND MISSING, CONGRUENT WITH t2
    o8: [] <- JE_1 [b2] (taken: false)
    o19: [] <- (out) []
b1 (exit, return, freq: 13):
    o23: [] <- (in) []
    o9: [t0] <- (copy) [t2]
    o10: [t3] <- MOV32ri [-10017]
    o11: [t7] <- (copy) [t0]
    o12: [t8] <- (copy) [t3]
    o13: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o26: [] <- (fun) [t7:edi,t8:esi] (call: o13)
    o20: [] <- (out) [rsp,edi,esi] <<< BOGUS OPERANDS
b2 (exit, return, freq: 8):
    o24: [] <- (in) []
    o15: [t4] <- MOV32r0 []
    o17: [] <- RETQ [42]
    o21: [] <- (out) [t4:eax]
```

_*.uni, output from enforceCallerSaved:_
```
b0 (entry, freq: 21):
    o0: [t9:rdi] <- (in) []
    o1: [t1] <- (copy) [t9]
    o2: [] <- ADJCALLSTACKDOWN64 [0,0]
    o3: [t5] <- (copy) [t1]
    o4: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o25: [t6:eax,t10:r004,t11:r005,t12:r006,t13:r007,t14:rcx,t15:rdx,t16:rsi,t17:rdi,t18:r8,t19:r9,t20:r10,t21:r11] <- (fun) [t5:rdi] (call: o4)
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [t2] <- (copy) [t6]
    o7: [] <- TEST32rr [t2,t2] <<< DEF OPERAND MISSING, CONGRUENT WITH t2
    o8: [] <- JE_1 [b2] (taken: false)
    o19: [] <- (out) []
b1 (exit, return, freq: 13):
    o23: [] <- (in) []
    o9: [t0] <- (copy) [t2]
    o10: [t3] <- MOV32ri [-10017]
    o11: [t7] <- (copy) [t0]
    o12: [t8] <- (copy) [t3]
    o13: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o26: [t22:rax,t23:rcx,t24:rdx,t25:rsi,t26:rdi,t27:r8,t28:r9,t29:r10,t30:r11] <- (fun) [t7:edi,t8:esi] (call: o13)
    o20: [] <- (out) [rsp,edi,esi] <<< BOGUS OPERANDS
b2 (exit, return, freq: 8):
    o24: [] <- (in) []
    o15: [t4] <- MOV32r0 []
    o17: [] <- RETQ [42]
    o21: [] <- (out) [t4:eax]
```

_*.uni, output from enforceCalleeSaved, reserveRegs, implementFrameOperations_
```
b0 (entry, freq: 21):
    o0: [t9:rdi,rbx,rbp,r12,r13,r14,r15,rsp] <- (in) []
    o1: [t1] <- (copy) [t9]
    o2: [] <- ADJCALLSTACKDOWN64 [0,0]
    o3: [t5] <- (copy) [t1]
    o4: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o25: [t6:eax,t10:r004,t11:r005,t12:r006,t13:r007,t14:rcx,t15:rdx,t16:rsi,t17:rdi,t18:r8,t19:r9,t20:r10,t21:r11] <- (fun) [t5:rdi] (call: o4)
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [t2] <- (copy) [t6]
    o7: [] <- TEST32rr [t2,t2] <<< DEF OPERAND MISSING, CONGRUENT WITH t2
    o8: [] <- JE_1 [b2] (taken: false)
    o19: [] <- (out) []
b1 (exit, return, freq: 13):
    o23: [] <- (in) []
    o9: [t0] <- (copy) [t2]
    o10: [t3] <- MOV32ri [-10017]
    o11: [t7] <- (copy) [t0]
    o12: [t8] <- (copy) [t3]
    o13: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o26: [t22:rax,t23:rcx,t24:rdx,t25:rsi,t26:rdi,t27:r8,t28:r9,t29:r10,t30:r11] <- (fun) [t7:edi,t8:esi] (call: o13)
    o20: [] <- (out) [rsp,edi,esi,rbx,rbp,r12,r13,r14,r15] <<< BOGUS edi, esi
b2 (exit, return, freq: 8):
    o24: [] <- (in) []
    o15: [t4] <- MOV32r0 []
    o17: [] <- RETQ [42]
    o21: [] <- (out) [t4:eax,rbx,rbp,r12,r13,r14,r15,rsp]
```

_*.uni, output from liftUndefRegs, killUnusedTemps,_ 
```
b0 (entry, freq: 21):
    o0: [t9:rdi,rbx,rbp,r12,r13,r14,r15,rsp,edi,esi] <- (in) [] <<< BOGUS edi, esi
    o1: [t1] <- (copy) [t9]
    o2: [] <- ADJCALLSTACKDOWN64 [0,0]
    o3: [t5] <- (copy) [t1]
    o4: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o25: [t6:eax,t10:r004,t11:r005,t12:r006,t13:r007,t14:rcx,t15:rdx,t16:rsi,t17:rdi,t18:r8,t19:r9,t20:r10,t21:r11] <- (fun) [t5:rdi] (call: o4)
    o27: [] <- (kill) [t10,t11,t12,t13,t14,t15,t16,t17,t18,t19,t20,t21]
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [t2] <- (copy) [t6]
    o7: [] <- TEST32rr [t2,t2] <<< DEF OPERAND MISSING, CONGRUENT WITH t2
    o8: [] <- JE_1 [b2] (taken: false)
    o19: [] <- (out) []
b1 (exit, return, freq: 13):
    o23: [] <- (in) []
    o9: [t0] <- (copy) [t2]
    o10: [t3] <- MOV32ri [-10017]
    o11: [t7] <- (copy) [t0]
    o12: [t8] <- (copy) [t3]
    o13: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o26: [t22:rax,t23:rcx,t24:rdx,t25:rsi,t26:rdi,t27:r8,t28:r9,t29:r10,t30:r11] <- (fun) [t7:edi,t8:esi] (call: o13)
    o28: [] <- (kill) [t22,t23,t24,t25,t26,t27,t28,t29,t30]
    o20: [] <- (out) [rsp,edi,esi,rbx,rbp,r12,r13,r14,r15] <<< BOGUS edi, esi
b2 (exit, return, freq: 8):
    o24: [] <- (in) []
    o15: [t4] <- MOV32r0 []
    o17: [] <- RETQ [42]
    o21: [] <- (out) [t4:eax,rbx,rbp,r12,r13,r14,r15,rsp]
```

_*.uni, output from extractRegs, foldCopies, splitBlocks, renameBlocks, repairCSSA, advancePhis, postponeBranches, renameTemps, sortGlobalTemps, renameOperations, normalizeFrequency, tagRemats, addPragmas_
```
b0 (entry, freq: 21):
    o0: [t0:rdi,t1:rbx,t2:rbp,t3:r12,t4:r13,t5:r14,t6:r15,t7:rsp,t8:edi,t9:esi] <- (in) [] <<< BOGUS edi, esi
    o1: [] <- ADJCALLSTACKDOWN64 [0,0]
    o2: [] <- CALL64pcrel32 [@pthread_mutex_unlock]
    o3: [t10:eax,t11:r004,t12:r005,t13:r006,t14:r007,t15:rcx,t16:rdx,t17:rsi,t18:rdi,t19:r8,t20:r9,t21:r10,t22:r11] <- (fun) [t0:rdi] (call: o2)
    o4: [] <- (kill) [t11,t12,t13,t14,t15,t16,t17,t18,t19,t20,t21,t22]
    o5: [] <- ADJCALLSTACKUP64 [0,0]
    o6: [] <- TEST32rr [t10,t10] <<< DEF OPERAND MISSING, CONGRUENT WITH t10
    o7: [] <- JE_1 [b2] (taken: false)
    o8: [] <- (out) []
b1 (exit, return, freq: 13):
    o9: [] <- (in) []
    o10: [t23] <- MOV32ri [-10017]
    o11: [] <- TCRETURNdi64 [@spio_map_unix_error,0]
    o12: [t24:rax,t25:rcx,t26:rdx,t27:rsi,t28:rdi,t29:r8,t30:r9,t31:r10,t32:r11] <- (fun) [t10:edi,t23:esi] (call: o11)
    o13: [] <- (kill) [t24,t25,t26,t27,t28,t29,t30,t31,t32]
    o14: [] <- (out) [t1:rbx,t2:rbp,t3:r12,t4:r13,t5:r14,t6:r15,t7:rsp,t8:edi,t9:esi] <<< BOGUS edi, esi
b2 (exit, return, freq: 8):
    o15: [] <- (in) []
    o16: [t33] <- MOV32r0 []
    o17: [] <- RETQ [42]
    o18: [] <- (out) [t1:rbx,t2:rbp,t3:r12,t4:r13,t5:r14,t6:r15,t7:rsp,t33:eax]
```

**Comment by matsc-at-sics-se** [comment]
Correction, the TEST32rr operations are fine, they define nothing.

**Comment by robcasloz** [comment]
A first step is to extend the target function `tInstructionType` to return `TailCallInstructionType` for `TCRETURNdi64`, see for example how `TCRETURNi` is handled in the Hexagon target.

**Comment by matsc-at-sics-se** [comment]
Fixed by commit fc63bfc.
